File: backend/src/main/java/com/github/denvned/xodus/compare/backend/graphql/ChangedEntityConnection.java
package com.github.denvned.xodus.compare.backend.graphql;

import com.github.denvned.graphql.annotations.GraphQLField;
import com.github.denvned.graphql.annotations.GraphQLNonNull;
import jetbrains.exodus.entitystore.EntityIterable;

import java.util.AbstractList;
import java.util.List;

public final class ChangedEntityConnection extends AbstractEntityBasedConnection {
    public ChangedEntityConnection(EntityIterable entities, Integer first, Long after) {
        super(entities, first, after);
    }

    @GraphQLField @GraphQLNonNull
    public List<@GraphQLNonNull ChangedEntityEdge> getEdges() {
        init();

        return new AbstractList<ChangedEntityEdge>() {
            @Override
            public ChangedEntityEdge get(int index) {
                return new ChangedEntityEdge(entities.get(index));
            }

            @Override
            public int size() {
                return entities.size();
            }
        };
    }
}

File: backend/src/main/kotlin/com/github/denvned/xodus/compare/backend/graphql/ChangedEntityConnection.kt
package com.github.denvned.xodus.compare.backend.graphql

import jetbrains.exodus.entitystore.EntityIterable
import java.util.*

class ChangedEntityConnection(entities: EntityIterable, first: Int?, after: Long?)
    : AbstractEntityBasedConnection(entities, first, after) {

  val edges: List<ChangedEntityEdge>
    get() = object : AbstractList<ChangedEntityEdge>() {
      override val size get() = page.entities.size
      override fun get(index: Int) = ChangedEntityEdge(page.entities[index])
    }
}
